src/service/video-manager-impl.js

```diff
@@ -0,0 +1,318 @@
+   * @return {boolean}
+   * @package
+   */
+  needsVisibilityObserver() {
+    return this.canAutoplay_;
+  }
+
+  /* Autoplay Behaviour */
+
+  /**
+   * Called when an autoplay video is built.
+   * @private
+   */
+  autoplayVideoBuilt_() {
+    this.video.mute();
+
+    // If autoplay video has controls, hide them and only show them on
+    // user-ineraction.
+    if (this.video.element.hasAttribute(VideoAttributes.CONTROLS)) {
+      this.video.hideControls();
+
+      // TODO(aghassemi): This won't work for iframes, needs a transparent shim
+      listenOnce(this.video.element, 'click', () => {
+        this.userInteracted_ = true;
+        this.video.showControls();
+        this.video.unmute();
+      });
+    }
+  }
+
+  /**
+   * Called when visibility of a loaded autoplay video changes.
+   * @private
+   */
+  autoplayLoadedVideoVisibilityChanged_() {
+    if (this.userInteracted_) {
+      return;
+    }
+
+    if (this.isVisible_) {
+      this.video.play(/*autoplay*/ true);
+    } else {
+      this.video.pause();
+    }
+  }
+
+  /**
+   * Called by all possible events that might change the visibility of the video
+   * such as scrolling or {@link ../video-interface.VideoEvents#VISIBILITY}.
+   * @package
+   */
+  updateVisibility() {
+    const wasVisible = this.isVisible_;
+
+    // Measure if video is now in viewport and what percentage of it is visible
+    const measure = () => {
+      if (!this.video.isInViewport()) {
+        this.isVisible_ = false;
+        return;
+      }
+
+      // Calculate what percentage of the video is in viewport
+      const change = this.video.element.getIntersectionChangeEntry();
+      const ir = change.intersectionRect;
+      const br = change.boundingClientRect;
+      const visiblePercent = br.height * br.width == 0 ? 0 :
+        ir.width * ir.height * 100 / (br.height * br.width);
+
+      this.isVisible_ = visiblePercent >= VISIBILITY_PERCENT;
+    };
+
+    // Mutate if visibility changed from previous state
+    const mutate = () => {
+      if (this.isVisible_ != wasVisible) {
+        this.videoVisibilityChanged_();
+      }
+    };
+
+    this.vsync_.run({
+      measure,
+      mutate,
+    });
+  }
+}
+
+/**
+ * Detects whether the platform even supports autoplay videos.
+ * @private
+ */
+function platformSupportsAutoplay(platform) {
+  // non-mobile platforms always supported autoplay
+  if (!platform.isAndroid() && !platform.isIos()) {
+    return true;
+  }
+
+  const version = platform.getMajorVersion();
+  if (platform.isAndroid()) {
+    if (platform.isFirefox() && version >= 37) {
+      return true;
+    }
+    if (platform.isChrome() && version >= 53) {
+      return true;
+    }
+  }
+
+  if (platform.isIos()) {
+    if (platform.isSafari() && version >= 10) {
+      return true;
+    }
+  }
+
+  // TODO(aghassemi): Test other combinations and add support.
+  // TODO(aghassemi): Is there a way to detect that autoplay has been disabled
```

Comment:
In preparation for that you may consider to return `!Promise<boolean>` instead `boolean`, so that a slightly async feature test can be done.